### DISCUSSION TITLE
Failed to initialize a runner; number value is required

### DISCUSSION THREAD

**Discussion by LinguineCode** [terraform-linters/tflint#1545] [Q&A] [answered]
## Introduction

Hello, I'm setting up `tflint` from a fresh start and I'm getting the error "Failed to initialize a runner; number value is required". Any idea what I'm doing wrong? Googling this error shows no relevant results so I'm hoping this post can help someone in the future.

## Debug output

```
$ TFLINT_LOG=debug tflint
13:30:58 config.go:126: [INFO] Load config: .tflint.hcl
13:30:58 config.go:136: [INFO] file not found
13:30:58 config.go:144: [INFO] Load config: /Users/myname/.tflint.hcl
13:30:58 config.go:152: [INFO] file not found
13:30:58 config.go:154: [INFO] Use default config
13:30:58 config.go:298: [INFO] The `terraform` plugin block is not found. Enable the plugin `terraform` automatically
13:30:58 option.go:49: [DEBUG] CLI Options
13:30:58 option.go:50: [DEBUG]   Module: false
13:30:58 option.go:51: [DEBUG]   Force: false
13:30:58 option.go:52: [DEBUG]   IgnoreModules:
13:30:58 option.go:56: [DEBUG]   EnableRules: 
13:30:58 option.go:57: [DEBUG]   DisableRules: 
13:30:58 option.go:58: [DEBUG]   Only: 
13:30:58 option.go:59: [DEBUG]   EnablePlugins: 
13:30:58 option.go:60: [DEBUG]   Varfiles: 
13:30:58 option.go:61: [DEBUG]   Variables: 
13:30:58 option.go:62: [DEBUG]   Format: 
13:30:58 loader.go:43: [INFO] Initialize new loader
13:30:58 loader.go:71: [INFO] Module inspection is disabled. Building a root module without children...
13:30:58 loader.go:121: [INFO] Load values files
13:30:58 runner.go:48: [INFO] Initialize new runner for root
Failed to initialize a runner; number value is required
```

```console
$ tflint -v
TFLint version 0.41.0
$ terraform -v
Terraform v1.3.2
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v4.34.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
```

**Comment by bendrucker** [comment]
Not reproducible with an empty module, albeit on darwin_arm64 instead of amd. A reproducible report is needed.

**Comment by gtirloni** [comment]
I can reproduce this with 0.41 but not with 0.40. Not sure why this was turned into a discussion topic.

I'll try to update this thread later today with more debugging info. So far I can reproduce it inside an Alpine 3.16 container with tflint 0.41... but not outside on a Fedora 36 system.

**Comment by gtirloni** [comment]
The issue seems to be when a string variable is passed to count.

Fails:

```
variable "myvar" {               
  default = "3"
}                               

resource "google_compute_instance" "default" {
  count = var.myvar
}
```

Works:

```
variable "myvar" {               
  default = 3
}                               

resource "google_compute_instance" "default" {
  count = var.myvar
}
```

**Comment by gtirloni** [comment]
Let me fully acknowledge the TF code is wrong but terraform 1.x seems very forgiving and does what is expected. We'll be fixing our code thanks for this error with tflint but I'm wondering what tflint does differently from terraform itself in this case.

**Comment by rquadling** [comment]
I've managed to replicate the issue.

The original code was:
```hcl
variable "redis_number_of_nodes" {
  description = "The number of cache nodes within the ElastiCache cluster. This number must be greater or equal 2 to enable automatic failover."
  type        = string
  default     = "1"
}

variable "redis_snapshot_retention_limit" {
  description = "The number of days for which automated backups are retained. Setting this parameter to a positive number enables backups. Setting this parameter to 0 disables automated backups"
  type        = string
  default     = "7"
}
```

The new code is:
```hcl
variable "redis_number_of_nodes" {
  description = "The number of cache nodes within the ElastiCache cluster. This number must be greater or equal 2 to enable automatic failover."
  type        = number
  default     = 1
}

variable "redis_snapshot_retention_limit" {
  description = "The number of days for which automated backups are retained. Setting this parameter to a positive number enables backups. Setting this parameter to 0 disables automated backups"
  type        = number
  default     = 7
}
```

The issue is that the type for the variable was wrong, and the default matched the defined type, not the type associated with the usage (which, in this case, was both numeric).

It would seem that strings, containing numbers, are treated as valid by Terraform (the really painful art of type juggling ... yeah!), but TFLint does the right thing (in my mind) and says "Nope" to them.

What would have been useful was to output SOMETHING about the actual error trigger.

**Reply by wata727** [reply]
As you say, Terraform casts the type to number from string in the `count` attribute. This is the Terraform language specification and TFLint should follow it. This seems to be a bug that occurred in v0.41.

**Reply by wata727** [reply]
This bug is tracked in https://github.com/terraform-linters/tflint/issues/1547.

**Comment by bendrucker** [comment, marked as answer]
## Summary

* Appreciate the reproductions
* #1547 tracks/explains the bug
* You can work around this now by ensuring that the values you supply to `count` are a `number`, which is the expected type
* Terraform accepts types that are convertible to `number` (`string`), that TFLint now doesn't is an unintended regression

**Reply by rquadling** [reply]
Can the error be improved to indicate what was being tested?
